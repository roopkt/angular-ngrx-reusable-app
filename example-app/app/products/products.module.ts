import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HomeContainerComponent } from './containers/home-container.component';

const COMPONENTS = [HomeContainerComponent];

@NgModule({
  imports: [
    CommonModule
  ],
  declarations: COMPONENTS,
  exports: COMPONENTS
})
export class WebProductModule { }
