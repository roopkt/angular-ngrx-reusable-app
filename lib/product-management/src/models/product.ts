export class Product {
    constructor(
        public id: string,
        public name: string) {
    }
}

export function generateMockProduct(): Product {
    return new Product('2', 'Apple');
}
