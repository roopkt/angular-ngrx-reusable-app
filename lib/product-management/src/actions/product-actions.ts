import { Action } from '@ngrx/store';
import { Product } from '../models/product';

export enum ProductActionTypes {
    ADD = '[Products] Add',
    LOAD = '[Products] Load',
}

export class Add implements Action {
    readonly type = ProductActionTypes.ADD;
    constructor(public payload: Product) { }
}

export class Load implements Action {
    readonly type = ProductActionTypes.LOAD;
    constructor() { }
}

export type ProductActions = Add | Load;



