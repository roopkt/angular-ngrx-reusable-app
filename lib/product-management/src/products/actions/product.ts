import { Action } from '@ngrx/store';
import { Product } from '../models/product';


export const ADD = '[Releases] Add';

export class Add implements Action {
    readonly type = ADD;
    constructor(public payload: Product) { }
}

export type Actions = Add;



