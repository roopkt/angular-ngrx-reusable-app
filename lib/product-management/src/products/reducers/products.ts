import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';
import * as productActions from '../actions/Product';
import { Product } from '../models/product';

export interface State extends EntityState<Product> {
    selectedProductId: string | null;
}

/**
 * createEntityAdapter creates many an object of helper
 * functions for single or multiple operations
 * against the dictionary of records. The configuration
 * object takes a record id selector function and
 * a sortComparer option which is set to a compare
 * function if the records are to be sorted.
 */
export const adapter: EntityAdapter<Product> = createEntityAdapter<Product>({
    selectId: (product: Product) => product.id,
    sortComparer: false,
});

/** getInitialState returns the default initial state
 * for the generated entity state. Initial state
 * additional properties can also be defined.
*/
export const initialState: State = adapter.getInitialState({
    selectedProductId: null,
});

export function reducer(state = initialState,
    action: productActions.Actions): State {
    switch (action.type) {
        case productActions.ADD: {
            return {
                ...adapter.addOne(action.payload, state),
                selectedProductId: state.selectedProductId
            };
        }
        default: {
            return state;
        }
    }
}

export const getSelectedId = (state: State) => state.selectedProductId;
