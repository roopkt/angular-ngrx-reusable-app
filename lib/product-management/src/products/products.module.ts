import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';

import { reducers } from './reducers';
import { AddProductContainerComponent } from './containers/add-product-container';
import { AddProductComponent } from './components/add-product';

const COMPONENTS = [
    AddProductComponent,
    AddProductContainerComponent
];

@NgModule({
    imports: [
        CommonModule,
        /**
         * StoreModule.forFeature is used for composing state
         * from feature modules. These modules can be loaded
         * eagerly or lazily and will be dynamically added to
         * the existing state.
         */
        StoreModule.forFeature('products', reducers),

        /**
         * Effects.forFeature is used to register effects
         * from feature modules. Effects can be loaded
         * eagerly or lazily and will be started immediately.
         *
         * All Effects will only be instantiated once regardless of
         * whether they are registered once or multiple times.
         */
        EffectsModule.forFeature([]),
    ],
    declarations: COMPONENTS,
    providers: [],
    exports: COMPONENTS
})
export class ProductsModule { }
