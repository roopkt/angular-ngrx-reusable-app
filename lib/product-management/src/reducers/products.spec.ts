import { reducer } from './products';
import * as fromProducts from './products';
import { Load, Add } from '../actions/product-actions';
import { Product, generateMockProduct } from '../models/product';


describe('ProductsReducer', () => {
    const product1 = generateMockProduct();
    const product2 = { ...product1, id: '222' };
    const product3 = { ...product1, id: '333' };
    const initialState: fromProducts.State = {
        ids: [product1.id, product2.id],
        entities: {
            [product1.id]: product1,
            [product2.id]: product2,
        },
        selectedProductId: null,
    };

    describe('undefined action', () => {
        it('should return the default state', () => {
            const result = reducer(undefined, {} as any);
            expect(result).toMatchSnapshot();
        });
    });

    describe('ADD', () => {
        const expectedResult = {
            ids: [product1.id],
            entities: {
                [product1.id]: product1,
            },
            selectedProductId: null
        };
        it('should add a single product', () => {
            const action = new Add(product1);
            const result = reducer(fromProducts.initialState, action);
            expect(result).toMatchSnapshot();
        });
    });
    describe('Selectors', () => {
        describe('getSelectedId', () => {
            it('should return the selected id', () => {
                const result = fromProducts.getSelectedId({
                    ...initialState,
                    selectedProductId: product1.id,
                });
                expect(result).toMatchSnapshot();
            });
        });
    });
});
