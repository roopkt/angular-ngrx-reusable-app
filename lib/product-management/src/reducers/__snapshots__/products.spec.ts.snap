// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsReducer ADD should add a single product 1`] = `
Object {
  "entities": Object {
    "2": Product {
      "id": "2",
      "name": "Apple",
    },
  },
  "ids": Array [
    "2",
  ],
  "selectedProductId": null,
}
`;

exports[`ProductsReducer Selectors getSelectedId should return the selected id 1`] = `"2"`;

exports[`ProductsReducer undefined action should return the default state 1`] = `
Object {
  "entities": Object {},
  "ids": Array [],
  "selectedProductId": null,
}
`;
